tempesta_fw/t/functional/very_many_backends/multi_backend.py

```diff
@@ -0,0 +1,91 @@
+    return struct.unpack("!L", packed)[0]
+
+def ip_number_to_str(ip_addr):
+    """ Convert ip in numeric form to string """
+    packed = struct.pack("!L", ip_addr)
+    return socket.inet_ntoa(packed)
+
+def create_interface(iface_id, base_interface_name, base_interface_ip):
+    """ Create interface alias for listeners on nginx machine """
+    base_ip = ip_str_to_number(base_interface_ip)
+    iface_ip_numeric = base_ip + iface_id
+    iface_ip = ip_number_to_str(iface_ip_numeric)
+    command = "LANG=C ifconfig %s:%i %s up" % (base_interface_name, iface_id, iface_ip)
+    remote.server.run_cmd(command)
+    return ("%s:%i" % (base_interface_name, iface_id), iface_ip)
+
+def remove_interface(iface_name):
+    """ Remove interface """
+    remote.server.run_cmd("LANG=C ifconfig %s down" % iface_name)
+
+class MultipleBackends(stress.StressTest):
+    """ Testing for 1M backends """
+    interface = "eth0"
+    base_ip = "192.168.10.1"
+    interfaces = []
+    config = 'cache 0;\n'
+
+    base_port = 16384
+
+    def create_interfaces(self, base_interface_name, base_interface_ip, number_of_ip):
+        """ Create specified amount of interface aliases """
+        for i in range(number_of_ip):
+            name = create_interface(i, base_interface_name, base_interface_ip)
+            self.interfaces.append(name)
+
+    def remove_interfaces(self):
+        """ Remove previously created interfaces """
+        if len(self.interfaces) > 0:
+            iface = self.interfaces[0]
+            iface_name = iface[0]
+            remove_interface(iface_name)
+
+        for iface in self.interfaces:
+            iface_name = iface[0]
+            self.interfaces.remove(iface)
+
+    def create_servers(self):
+        self.interface = tf_cfg.cfg.get('Server', 'listen_interface')
+        self.create_interfaces(self.interface, self.base_ip, self.num_interfaces)
+        for iface in self.interfaces:
+            ip_addr = iface[1]
+            server = control.Nginx(start_listen_port=self.base_port, \
+                                   listen_ports=self.num_listeners_per_interface, \
+                                   listen_ip=ip_addr)
+            self.servers.append(server)
+
+    def configure_tempesta(self):
+        """ Configure tempesta 1 port in group """
+        sgid = 0
+        for server in self.servers:
+            for listener in server.config.listeners:
+                server_group = tempesta.ServerGroup('default-%i' % sgid)
+                server_group.add_server(server.ip, listener.port, server.conns_n)
+                self.tempesta.config.add_sg(server_group)
+                sgid = sgid + 1
+
+    def tearDown(self):
+        """ Stop nginx and tempesta, clear interfaces after this """
+        super(MultipleBackends, self).tearDown()
+        tf_cfg.dbg(2, "Removing interfaces")
+        self.remove_interfaces()
+
+    def test(self):
+        """ Test 1M backends """
+        self.generic_test_routine(self.config)
```

Comment:
It seems you don't test how long does it take to add a 1,000,001'st server. The test must be implemented as well.

Reply:
fixed